wgengine/magicsock/magicsock.go

```diff
@@ -2599,86 +2595,91 @@ func (c *Conn) initialBind() error {
+	ctx := context.Background() // unused without DNS name to resolve
 	addr := net.JoinHostPort(host, fmt.Sprint(port))
 	if c.packetListener != nil {
 		return c.packetListener.ListenPacket(ctx, network, addr)
 	}
 	return netns.Listener().ListenPacket(ctx, network, addr)
 }
 
-func (c *Conn) bind1(ruc **RebindingUDPConn, which string) error {
+// bind1 binds a UDP socket.
+// If there was an existing UDP socket bound, it closes it.
+// The caller is responsible for informing the portMapper of any changes.
+func (c *Conn) bind1(rucPtr **RebindingUDPConn, which string) error {
 	host := ""
 	if inTest() && !c.simulatedNetwork {
 		host = "127.0.0.1"
 		if which == "udp6" {
 			host = "::1"
 		}
 	}
-	var pc net.PacketConn
-	var err error
-	listenCtx := context.Background() // unused without DNS name to resolve
-	if c.port == 0 && DefaultPort != 0 {
-		pc, err = c.listenPacket(listenCtx, which, host, DefaultPort)
-		if err != nil {
-			c.logf("magicsock: bind: default port %s/%v unavailable; picking random", which, DefaultPort)
-		}
+
+	if *rucPtr == nil {
+		*rucPtr = new(RebindingUDPConn)
 	}
-	if pc == nil {
-		pc, err = c.listenPacket(listenCtx, which, host, c.port)
+	ruc := *rucPtr
+
+	// Hold the ruc lock the entire time, so that the close+bind is atomic
+	// from the perspective of ruc receive functions.
+	ruc.mu.Lock()
+	defer ruc.mu.Unlock()
+
+	// Build a list of preferred ports.
+	// Best is the port that the user requested.
+	// Second best is the port that is currently in use.
+	// If those fail, fall back to 0.
+	// The list of preferred ports might contain duplicates.
+	// Slightly wasteful, but no real harm to that.
+	var ports []uint16
+	if c.port != 0 {
+		ports = append(ports, c.port)
 	}
-	if err != nil {
-		c.logf("magicsock: bind(%s/%v): %v", which, c.port, err)
-		return fmt.Errorf("magicsock: bind: %s/%d: %v", which, c.port, err)
+	if ruc.pconn != nil {
+		curPort := uint16(ruc.localAddrLocked().Port)
+		ports = append(ports, curPort)
 	}
-	if *ruc == nil {
-		*ruc = new(RebindingUDPConn)
+	ports = append(ports, 0)
+
+	var pconn net.PacketConn
+	for _, port := range ports {
+		// Close the existing conn, in case it is sitting on the port we want.
+		err := ruc.closeLocked()
+		if err != nil && !errors.Is(err, net.ErrClosed) && !errors.Is(err, errNilPConn) {
+			c.logf("magicsock: bind1 %v close failed: %v", which, err)
+		}
+		// Open a new one with the desired port.
+		pconn, err = c.listenPacket(which, host, port)
+		if err != nil {
+			c.logf("magicsock: unable to bind %v port %d: %v", which, port, err)
+			continue
+		}
+		// Success.
+		ruc.pconn = pconn
+		return nil
 	}
-	(*ruc).Reset(pc)
-	return nil
+
+	// Failed to bind, including on port 0 (!).
+	// Set pconn to a dummy conn whose reads block until closed.
+	// This keeps the receive funcs alive for a future in which
+	// we get a link change and we can try binding again.
+	ruc.pconn = newBlockForeverConn()
```

Comment:
you going to wire this into the health package? to make sure we don't get stuck in this state.

at leat a TODO for now? maybe a bug?

Reply:
Will do in a follow-up.